protocol.md

```diff
@@ -870,6 +902,12 @@ export interface TextDocumentClientCapabilities {
 ```typescript
 interface ClientCapabilities {
 	/**
+	 * The client supports receiving the result solely through $/partialResult notifications for requests from the client to the server.
+	 * If true, the server should send all data through $/partialResult and send null as the result in the final response.
```

Comment:
> We are talking about two different clients. When I said client, I mean the consumer of the LSP wire format (i.e. vscode-languageclient). vscode-languageclient will have to support both streaming and non-streaming. Higher level APIs can be designed however is most convenient regardless of wire protocol.

I don't think we are talking about different clients - I think the vscode-languageclient could easily transform non-streamed responses into streamed ones and transform streamed responses into non-streamed. It could even provide and support both APIs without breaking changes.

> Hmm, yeah this does seem like a problem. JSON RPC doesn't mention or define any sort of ordering guarantees between different requests or notifications. Even if all requests are streaming, this is a critical problem for the ordering of notifications (i.e. $/partialResult).

It is true that the JSON RPC spec doesn't say anywhere that messages must be ordered, but it doesn't say anywhere that it is allowed to mangle them either. And if you think about this, having them ordered is the only way that makes sense and that's probably the reason they didn't spec this, it's the only logical way.
JSON RPC is a remote _procedure call protocol_. If you call two two arbitrary methods that can have arbitrary side effects, of course the ordering matters. If the client says call A, then B, but the JSON RPC library will sometimes tell the server the client called B, then A, that is a critical bug in the JSON RPC library.
And LSP _already_ relies on this. For example, you cannot send any request before you send `initialize`. If the JSON RPC library can in theory mess up the ordering, then a correct sequence of `initialize` + other method could in theory result in other method being called before `initialize`, which is wrong. Another example: A user opens a file, then closes it. `textDocument/open` causes the LS to load the content into memory, `textDocument/close` to remove it from memory. Both are notifications. Now if the user does this very quickly and the JSON RPC library does not guarantee ordering because of race conditions, this could result in the server first "closing" the not-opened document (a noop or silent error) and then opening it, and now you have a text document in memory that will never be closed.

I think you are overthinking this issue on the protocol level. This is something that needs to be fixed in the library.